homeassistant/components/gardena_bluetooth/__init__.py

```diff
@@ -0,0 +1,89 @@
+"""The Gardena Bluetooth integration."""
+from __future__ import annotations
+
+import asyncio
+from datetime import datetime
+import logging
+
+from bleak.exc import BleakError
+from gardena_bluetooth import (
+    get_all_characteristics_uuid,
+    read_char,
+    update_timestamp,
+)
+from gardena_bluetooth.const import DeviceConfiguration, DeviceInformation
+from gardena_bluetooth.exceptions import CharacteristicNoAccess
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr
+import homeassistant.util.dt as dt_util
+
+from .const import DOMAIN
+from .coordinator import Coordinator, DeviceUnavailable
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+LOGGER = logging.getLogger(__name__)
+TIMEOUT = 20.0
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Gardena Bluetooth from a config entry."""
+
+    address = entry.data[CONF_ADDRESS]
+    device_registry = dr.async_get(hass)
+    coordinator = Coordinator(hass, LOGGER, address)
+
+    try:
+        async with coordinator.client() as client:
+            sw_version = await read_char(client, DeviceInformation.firmware_version)
+            manufacturer = await read_char(client, DeviceInformation.manufacturer_name)
```

Comment:
These feels like it's the equivalent of an http get.  I think it should live in the external lib

Reply:
They are parsed and turned into python objects in the lib. They could also have been entities, which would mean about the equivalent code. Moving it to lib would also require lib to decide what information is of interest, so im a bit hesitant.

Doing sw_version = dataobj.sw_version, vs a single line function call to get the data, is not much benefit.

Reply:
I've reworked it somewhat. Moved the default handling into lib. But i feel like only HA known what fields are of interest for device information.

Reply:
Marking as resolved. It's wrapped up in a client lib instead of the bleak client now